orttraining/orttraining/training_api/optimizer.cc

```diff
@@ -5,58 +5,106 @@
+        inputs_.push_back(param_named_optimizer_states.at(param_name).momentum_named_states.at(utils::MOMENT_STATE_NAMES[0]));
+      } else if (utils::GetParamNameFromSuffix(name, utils::MOMENT_2_SUFFIX, param_name)) {
+        moment2_names.push_back(name);
+        auto it = named_parameters_.find(param_name);
+        ORT_ENFORCE(it != named_parameters_.end(), "Unknown param: ", param_name, " for field: ", name);
+        inputs_.push_back(param_named_optimizer_states.at(param_name).momentum_named_states.at(utils::MOMENT_STATE_NAMES[1]));
+      } else {
+        ORT_ENFORCE("This is an invalid graph. Optimizer graph contains unknown user input:", name);
+      }
+    }
+  }
+  // Add other optimizer reordering logic here
+  return Status::OK();
+}
 
 Optimizer::Optimizer(const std::string& optim_path_or_bytes,
-                     const std::unordered_map<std::string, std::shared_ptr<Parameter>>& parameters) {
-  std::unordered_map<std::string, ParameterOptimizerState>&
-      param_named_optimizer_states = optimizer_state_.param_named_optimizer_states;
-
+                     const std::unordered_map<std::string, std::shared_ptr<Parameter>>& parameters) : named_parameters_(parameters) {
+  // TODO: share threadpool, env with module session
   const SessionOptions session_options;
   std::unique_ptr<Environment> env;
   ORT_ENFORCE(Environment::Create(nullptr, env) == Status::OK(), "Enviroment creation fails.");
   optim_sess_ = std::move(std::make_unique<InferenceSession>(session_options, *env));
 
-  ORT_ENFORCE(optim_sess_->Load(optim_path_or_bytes).IsOK());
-  ORT_ENFORCE(optim_sess_->Initialize().IsOK());
+  ORT_THROW_IF_ERROR(optim_sess_->Load(optim_path_or_bytes));
+  ORT_THROW_IF_ERROR(optim_sess_->Initialize());
 
-  // TODO: don't hard code the state names, should get the state names according to the optimizer types.
-  std::vector<std::string> state_names{"momentum0", "momentum1"};
-  for (auto& pair : parameters) {
-    if (pair.second->RequiresGrad()) {
-      param_named_optimizer_states.insert({pair.first, ParameterOptimizerState()});
-      ParameterOptimizerState& cur_param_optimizer_states = param_named_optimizer_states[pair.first];
-      for (auto& state_name : state_names) {
-        OrtValue param_state;
-        // TODO: should reset the state to zero (for both CPU or CUDA Tensors.)
-        ORT_ENFORCE(CreateOrtValueFromOrtValue(pair.second->Data(), param_state, optim_sess_.get()).IsOK());
-        cur_param_optimizer_states.momentum_named_states.insert({state_name, std::make_shared<OrtValue>(param_state)});
-      }
-    }
+  utils::GetGraphInputOutputNames(optim_sess_, input_names_, output_names_);
+  ORT_ENFORCE(input_names_[0] == "learning_rate");  // TODO: make this better
+  ORT_ENFORCE(input_names_[1] == "step");           // TODO: make this better
+
+  if (optimizer_type_ == OptimizerType::AdamW || optimizer_type_ == OptimizerType::Lamb) {
```

Comment:
shall we only add the opt once we need it?

Reply:
removing lamb for now